libraries/Microsoft.Bot.Builder.AI.QnA/DialogHelper.cs

```diff
@@ -0,0 +1,188 @@
+        /// </summary>
+        public const string ActiveLearningDialogName = "active-learning-dialog";
+
+        /// <summary>
+        /// Gets QnA Maker Active Learning Dialog.
+        /// </summary>
+        public WaterfallDialog QnAMakerActiveLearningDialog { get; }
+        
+        // Define value names for values tracked inside the dialogs.
+        private const string CurrentQuery = "value-current-query";
+        private const string QnAData = "value-qnaData";
+
+        // Dialog Options parameters
+        private const float DefaultThreshold = 0.03F;
+        private const int DefaultTopN = 3;
+
+        // Card parameters
+        private const string CardTitle = "Did you mean:";
+        private const string CardNoMatchText = "None of the above.";
+        private const string CardNoMatchResponse = "Thanks for the feedback.";
+
+        private readonly QnAMaker _services;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DialogHelper"/> class.
+        /// Dialog helper to generate dialogs.
+        /// </summary>
+        /// <param name="services">Bot Services.</param>
+        public DialogHelper(QnAMaker services)
+        {
+            QnAMakerActiveLearningDialog = new WaterfallDialog(ActiveLearningDialogName)
+                .AddStep(CallGenerateAnswer)
+                .AddStep(CallTrain)
+                .AddStep(DisplayQnAResult);
+            _services = services;
+        }
+
+        private async Task<DialogTurnResult> CallGenerateAnswer(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            var scoreThreshold = DefaultThreshold;
+            var top = DefaultTopN;
+
+            QnAMakerOptions qnaMakerOptions = null;
+
+            // Getting options
+            if (stepContext.ActiveDialog.State["options"] != null)
+            {
+                qnaMakerOptions = stepContext.ActiveDialog.State["options"] as QnAMakerOptions;
+                scoreThreshold = qnaMakerOptions?.ScoreThreshold != null ? qnaMakerOptions.ScoreThreshold : DefaultThreshold;
+                top = qnaMakerOptions?.Top != null ? qnaMakerOptions.Top : DefaultTopN;
+            }
+            
+            // Storing the context info
+            stepContext.Values[CurrentQuery] = stepContext.Context.Activity.Text;
+            // -Check if previous context is present, if yes then put it with the query
+            // -Check for id if query is present in reverse index
+
+            var response = await _services.GetAnswersAsync(stepContext.Context, qnaMakerOptions).ConfigureAwait(false);
+
+            // TODO: Take this value from GetAnswerResponse 
+            var isActiveLearningEnabled = true;
+            var filteredResponse = response;
+
+            if (isActiveLearningEnabled)
+            {
+                filteredResponse = _services.GetLowScoreVariation(response.ToArray());
+
+                if (filteredResponse.ToList().Count > 1)
```

Comment:
Efficiency: Line 84 is potentially doing a copy of the response (I don't knwo what the return type is so I can't tell). 

Further more, this line 86, is doing another copy with ToList(), only to generate the count and throw away. Consider using Count(), and also consider not converting so much across different collection types.

Reply:
Please review this, I would prefer to not merge this as-is